src/interpreter/thunk.ts

```diff
@@ -0,0 +1,83 @@
+import * as es from 'estree'
+import { Context, Value } from '../types'
+
+export type EvaluateFunction = (node: es.Node, context: Context) => Generator<any>
+
+function getContextWithIndependentEnvironment(context: Context): Context {
+  const result = {
+    ...context,
+    runtime: {
+      ...context.runtime,
+      environments: [...context.runtime.environments]
+    }
+  }
+  return result
+}
+
+export default class Thunk {
+  public isEvaluated: boolean
+  public result: Value
+  public originalNode: es.Node
+  public context: Context
+  public evaluate: EvaluateFunction
```

Comment:
Interesting that the function to evaluate the thunk is stored in it. But if you have the original node, why would you need this function instead of the normal eager evaluate()?

Reply:
You are right. I tried to decouple the modules by using OOP. However, I realize that it created a lot of troubles. So I decided to revert to procedural programming.